group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroupTest.java

```diff
@@ -0,0 +1,500 @@
+
+        assignor = consumerGroup.preferredServerAssignor(member1, updatedMember1);
+        assertTrue(assignor.equals(Optional.of("range")) || assignor.equals(Optional.of("uniform")));
+
+        assertEquals(
+            Optional.of("uniform"),
+            consumerGroup.preferredServerAssignor(member2, updatedMember2)
+        );
+
+        consumerGroup.updateMember(updatedMember2);
+
+        assertEquals(
+            Optional.of("uniform"),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+
+        assertEquals(
+            Optional.empty(),
+            consumerGroup.preferredServerAssignor(member3, updatedMember3)
+        );
+
+        consumerGroup.updateMember(updatedMember3);
+
+        assertEquals(
+            Optional.empty(),
+            consumerGroup.preferredServerAssignor(null, null)
+        );
+    }
+
+    @Test
+    public void testUpdateSubscriptionMetadata() {
+        Uuid fooTopicId = Uuid.randomUuid();
+        Uuid barTopicId = Uuid.randomUuid();
+        Uuid zarTopicId = Uuid.randomUuid();
+
+        TopicsImage image = new GroupMetadataManagerTest.TopicsImageBuilder()
+            .addTopic(fooTopicId, "foo", 1)
+            .addTopic(barTopicId, "bar", 2)
+            .addTopic(zarTopicId, "zar", 3)
+            .build();
+
+        ConsumerGroupMember member1 = new ConsumerGroupMember.Builder("member1")
+            .setSubscribedTopicNames(Arrays.asList("foo"))
+            .build();
+        ConsumerGroupMember member2 = new ConsumerGroupMember.Builder("member2")
+            .setSubscribedTopicNames(Arrays.asList("bar"))
+            .build();
+        ConsumerGroupMember member3 = new ConsumerGroupMember.Builder("member3")
+            .setSubscribedTopicNames(Arrays.asList("zar"))
+            .build();
+
+        ConsumerGroup consumerGroup = createConsumerGroup("group-foo");
+
+        // It should be empty by default.
+        assertEquals(
+            Collections.emptyMap(),
+            consumerGroup.computeSubscriptionMetadata(
+                null,
+                null,
+                image
+            )
+        );
+
+        // Adding member1.
+        assertEquals(
+            mkMap(
+                mkEntry("foo", new TopicMetadata(fooTopicId, "foo", 1))
+            ),
+            consumerGroup.computeSubscriptionMetadata(
+                null,
+                member1,
+                image
+            )
+        );
+
+        // Updating the group with member1.
+        consumerGroup.updateMember(member1);
+
+        // It should return foo now.
+        assertEquals(
+            mkMap(
+                mkEntry("foo", new TopicMetadata(fooTopicId, "foo", 1))
+            ),
+            consumerGroup.computeSubscriptionMetadata(
+                null,
+                null,
+                image
+            )
+        );
+
+        // Removing member1 results in an empty map.
+        assertEquals(
```

Comment:
We aren't actually removing member1 here but computing what would happen if we did. This is a bit confusing to follow.

Reply:
I have put more comments. I hope it helps.

Reply:
This and the server assignor test are much more readable. Thanks!